### ISSUE TITLE
[feature] Change parameter `--format json` when producing pkglists or graphs

### ISSUE THREAD

**Issue by lhocmf** [conan-io/conan#15883]
### What is your suggestion?

# Multiple formats
<details><summary>Currently available JSON formats</summary>

As far as I can tell, there are currently these JSON formats in use:

* lockfile
   * output by `--lockfile-out <file>`
      * `conan graph build-order`
      * `conan install`
      * `conan lock [add|create|merge|remove]`
      * `conan build`
      * `conan create`
      * `conan export`
      * `conan export-pkg`
   * input via `--lockfile <file>`
      * `conan graph build-order`
      * `conan inspect`
      * `conan install`
      * `conan lock [add|merge|remove]`
      * `conan build`
      * `conan create`
      * `conan export`
      * `conan export-pkg`
* graph info
   * output by:
      * `conan graph build-order[-merge] --format json`
      * `conan graph explain --format json`
      * `conan graph info --format json`
      * `conan install --format json`
      * `conan build --format json`
      * `conan create --format json`
      * `conan export-pkg --format json`
   * input of:
      * `conan list --graph <file>`
      * `conan graph build-order-merge --file <file>`
* pkglist
   * output by:
      * `conan cache [save|restore] --format json`
      * `conan list --format json`
      * `conan remove --format json`
      * `conan search --format json`
      * `conan download --format json`
      * `conan export --format pkglist`
      * `conan upload --format json`
   * input of:
      * `conan download --list <file>`
      * `conan upload --list <file>`
      * `conan remove --list <file>`
      * `conan cache save --list <file>`
* some minor formats only used as output, e.g.
   * remotes (`conan remote [list|enable|disable] --format json`)
   * remote users (`conan remote [auth|list-users|login|logout|set-user] --format json`)
   * configuration (`conan config [list|show] --format json`)
   * version info (`conan version --format json`)
   * inspection (`conan inspect --format json`)
   * profiles (`conan profile list --format json`)
   * editables (`conan editable list --format json`)
   * recipe name/version (`conan export --format json`)

</details> 

There are several different JSON formats used by conan commands to parse data for input or output in a machine-parsable fashion. However, without looking deep into the documentation, it is not always clear which command accepts and/or produces which format. (Worse yet, even the documentation usually doesn't go into much detail of the structure of any given format. But I guess this is a topic for another discussion...)

For lockfiles, this is almost impossible to misinterpret, due to consistent usage of `--lockfile` and `--lockfile-out`.

For input files, this is almost consistent: All commands acception pkglists use a `--list` parameter, and graph files have a similar `--graph` parameter (`conan graph build-order-merge` as the only other command to consume graph files has `--file` instead).

However, for all graph and pkglist outputs, there usually is only `--format json`, independent of which format is actually produced (because they are all JSON?). I have found only one exception thus far: `conan export` has `--format json` and `--format pkglist`. (I guess this is likely because the `--format json` output format predates the pkglist format and needs to be preserved for compatibility).

This makes it hard to see, whether an output of one command can be accepted by another command.

# Example to highlight the problem

```
conan create --format json [...] >file.json
conan upload --list file.json [...]
```

Is this a valid command sequence? Hard to tell. The only hint to any "format" in the commands is the `--list` parameter, but that is only on the consumer, not the producer.

Consulting `conan create --help` and `conan list --help` doesn't help much, as the relevant option descriptions read `Select the output format: json` and `Package list file` respectively, giving no further information.
Consulting the documentation at https://docs.conan.io *might* give us a hint, that is confirmed by actually running the command sequence; No, this is not valid. The format produced by `conan create --format json` is not consumable by `conan upload --list <file>`, as `conan create --format json` produces graph info json.

A correct sequence for this scenario could be:
```
conan create --format json [...] >file.json
conan list --graph file.json --format json >file2.json
conan upload --list file2.json [...]
```

But even here, it is hard to tell whether `file2.json` could be used as input for `conan list --graph file2.json`, or whether `file.json` and `file2.json` are different formats at all.

# Solution 1: Name the formats explicitly

A simple change to the format name could improve this situation:
```
conan create --format graph [...] >file.json
conan list --graph file.json --format pkglist >file2.json
conan upload --list file2.json [...]
```

Now, there is a clear distinction between `--format graph`/`--graph <file>` and `--format pkglist`/`--list <file>` parameter pairs. (And there seems to be some kind of precedent with `conan export` already...)

I'm not sure whether changing `--format json` to `--format pkglist`/`--format graph` outright is feasible due to compatibility reasons, but could they maybe be added as aliases? That way, scripts designed against the current CLI continue working, while newer scripts could use the more descriptive format specifiers.

Then again, it is not clear to me from the documentation whether only the graph info format is experimental or the whole feature.

# Solution 2: Copy lockfile options

Similar to `--lockfile` and `--lockfile-out`, similar counterparts to `--graph` and `--list` could be added as `--graph-out` and `--list-out` respectively. This would look like this for above example:
```
conan create --graph-out file.json [...]
conan list --graph file.json --list-out file2.json
conan upload --list file2.json [...]
```

I personally would prefer this approach, as it mirrors the existing lockfile parameters. Additionally, it would allow the commands to produce other output (e.g. `conan export --format json --list-out list.json >export.json`) if required.

I would also argue that at least some of the commands producing graph info (`create`, `install`, `build`, `export-pkg`) or pkglist (`cache save`, `cache restore`, `download`, `upload`, `export`) do not produce these files as their primary purpose. Rather, the formatter is currently misused to generate these by-products in order to facilitate an additional communication channel by sacrifing a different one. Having a different parameter instead of `--format` would solve this conundrum as well. (And I can still capture `conan create` output AND  additinally get a graph info JSON file!).

But, it would be a more drastic change to current command line parameters. So I'm not quite sure whether this is feasible at the current maturity of these features.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @lhocmf 

Thanks for your suggestion.

We had this conversation while designing the CLI, and we finally leaned towards to having a more homogenous interface, using always ``--format=json`` for the structured format in json, even if those jsons can represent different concepts. As often, it is just a trade-off, we understand that the more "semantic" ``--format`` output would make more sense for some people, while others prefer the pure data format as specifier of the output. It is also true that we had to introduce at some point the ``conan export --format=pkglist``, and it was mainly done for one very important reason for us: not breaking existing users. If this would have been known 1 year ago, the ``--format=pkglist`` would not exist.

And this leads me to the fact that Conan 2 is already more than 1 year ago, and this change cannot be done, it would be a breaking change, affecting tons of users. This is against the Conan 2 stability commitment. It might be possible to introduce it gradually, make it co-exist, then deprecate and finally remove, but we have been discussing this within the team and honestly it would be too big of an effort, for something that we are not even fully convinced it would be the best UX for everyone.

I'd suggest to close this as not in scope for Conan 2, and we will take this feedback into account when designing an eventual Conan 3 version. Thanks very much for your feedback again!

**Comment by lhocmf** [comment]
@memsharded

Thanks for your reply. I feared this might be a hard obstacle to overcome, so I tried to limit my request to be as small a change as possible. I had hopes because the features affected are still documented as being experimental, but it seems even experimental features have to have some guarantee of stability (which I do like in other contexts).

I found this mostly because I had to refactor a part of our CI scripts that would have wanted to use both `--format json` and `--format pkglist` for `conan export` at the same time, and it seemed to me that these "outputs" should be independent of each other (as both return different results). That led me down the rabbit hole of "which `conan` command can read the output of which other `conan` commands... And here we are.
